File: python/1d-dynamic-programming/palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        count = 0
        for i in range(len(s)):
            distance, current = 0, s[i:i + 1]
            while current == current[::-1]:
                if i - distance < 0 or i + distance == len(s):
                    break
                distance += 1
                current = s[i - distance:i + distance + 1]
                count += 1

        for i in range(len(s) - 1):
            distance, current = 0, s[i:i + 2]
            while current == current[::-1]:
                if i - distance < 0 or i + 1 + distance >= len(s):
                    break
                distance += 1
                current = s[i - distance:i + distance + 2]
                count += 1

        return count

File: c++/1d-dynamic-programming/palindromic-substrings.cc
class Solution {
public:
    int countSubstrings(string s) {
        int palindromes = 0;

        for (int i = 0; i < s.size(); ++i) {
            int size = 0;
            while (i - size >= 0 && i + size < s.size() && s[i - size] == s[i + size]) {
                palindromes += 1;
                size += 1;
            }

            size = 0;
            while (i - size >= 0 && i + size + 1 < s.size() && s[i - size] == s[i + size + 1]) {
                palindromes += 1;
                size += 1;
            }
        }

        return palindromes;
    }
};
